### PR TITLE
[Merged by Bors] - feat: port Data.Nat.Nth

### PR DESCRIPTION
**Pull request by urkud**
---

- [x] depends on: #leanprover-community/mathlib#18760
- [x] depends on: #5034 
- [x] depends on: #5257

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by urkud** [comment]
@mo271 Are you working on this now? If no, then I'm going to fix the rest.

**Comment by mo271** [comment]
> @mo271 Are you working on this now? If no, then I'm going to fix the rest.

@urkud go ahead, I'm not working on it right now..

**Comment by urkud** [comment]
I fixed compile but I failed to stop myself from reviewing proofs/API. Now I'm going to split this PR into easily reviewable chunks.

**Comment by YaelDillies** [comment]
Could you update the SHA?

**Comment by eric-wieser** [comment]
To what? We're still waiting for #18760

**Comment by YaelDillies** [comment]
Not anymore

**Comment by urkud** [comment]
@YaelDillies It makes no sense to update this branch because I'm going to restart it once leanprover-community/mathlib#18760 is merged.

File: Mathlib/Data/Set/Finite.lean [hunk 1]
```diff
@@ -4,7 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Johannes Hölzl, Mario Carneiro, Kyle Miller
 
 ! This file was ported from Lean 3 source module data.set.finite
-! leanprover-community/mathlib commit 5bb9fffd23f9f65b367f5d451da18cc60bf47335
+! leanprover-community/mathlib commit 7fdd4f3746cb059edfdb5d52cba98f66fce418c0
```
Comment (eric-wieser) [inline Mathlib/Data/Set/Finite.lean:7]: Can you split this to its own PR?

Comment (urkud) [reply Mathlib/Data/Set/Finite.lean:7]: It is in #5034. I rebased this branch on top of that PR to remove it from the "out of sync" list.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#5034
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by urkud** [comment]
I'm going to rebase today. Please don't merge master.

File: Mathlib/Data/Nat/Nth.lean [hunk 2]
```diff
@@ -243,22 +242,23 @@ theorem isLeast_nth_of_infinite (hf : (setOf p).Infinite) (n : ℕ) :
...
+nonempty because we use the same "garbage value" `0` both for `Inf` on `ℕ` and for `Nat.nth s n` for
 `n ≥ card s`. -/
 theorem nth_eq_sInf (p : ℕ → Prop) (n : ℕ) : nth p n = sInf {x | p x ∧ ∀ k < n, nth p k < x} := by
-  by_cases hn : ∀ hf : (setOf p).Finite, n < hf.to_finset.card
-  · exact (is_least_nth hn).csInf_eq.symm
-  · push_neg at hn 
+  by_cases hn : ∀ hf : (setOf p).Finite, n < hf.toFinset.card
+  · exact (isLeast_nth hn).csInf_eq.symm
+  · -- porting note: TODO: was push_neg at hn
+    simp only [not_forall, not_lt] at hn
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Nat/Nth.lean:253]: `push_neg` should work better now

**Review by Ruben-VandeVelde — COMMENTED** [review]
Please check if you can restore `push_neg`; lgtm otherwise. Could you mark as awaiting-review if you agree?

File: Mathlib/Data/Nat/Nth.lean (hunk 2 above)
Comment (Komyyy) [reply Mathlib/Data/Nat/Nth.lean:253]: Because `push_neg at hn` makes a loose bounded variable.
I'm going to fix this bug.

Comment (Komyyy) [reply Mathlib/Data/Nat/Nth.lean]: #5257

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#5034~~
* ~~leanprover-community/mathlib4#5257~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**